### PR TITLE
3rd party services isolation/8

### PR DESCRIPTION
**Pull request by mikicho**
https://github.com/testjavascript/integration-tests-a-z/issues/8

Just not sure why we need this nor how to accomplish it:
> Show how to revert to default after eachTest so a test won't leave a dirty state

➿ this icon should be an "owl" (😄), for insightful comments which aren't best practices

### CODE REVIEW

**Comment by goldbergyoni** [comment]
@mikicho Working toward some big milestone on Sunday afternoon, plan to review by EOD Sunday

File: example-application/libraries/mailer.js [hunk 1]
```diff
@@ -1,3 +1,6 @@
-module.exports.send = (subject, body, recipientAddress) => {
+const axios = require('axios');
+
+module.exports.send = async (subject, body, recipientAddress) => {
   console.log('Not really a mailer, right?', subject, body, recipientAddress);
+  await axios.post(`http://localhost/mailer`, { subject, body, recipientAddress });
```
Comment (goldbergyoni) [inline example-application/libraries/mailer.js:5]: See the comment 'Intercept vs stub comment' below

We have to decide whether mailer is our code or npm lib. If the later, we should not intercept the URL

File: various-receipes/3rd-party-services-isolation/test/isolation.test.js [hunk 2]
```diff
@@ -0,0 +1,110 @@
+const request = require("supertest");
```
Comment (goldbergyoni) [inline various-receipes/3rd-party-services-isolation/test/isolation.test.js:1]: What is the value in this test vs the basic examples? If we dedicated a recipe for isolation, maybe focus on simulation and show how to react to 3rd party corner cases: unusual response, errors, network down (!), etc

**Review by goldbergyoni — CHANGES_REQUESTED** [review]
@mikicho Interesting topic you picked here:)

I'm not sure that mailer is a good example for isolation and simulation, maybe the user Microservice is as it will allow showing various behaviour like errors, various corner case responses, etc

Also, consider showing a typical pattern where there is a default behaviour (all calls to /user/:id/) returns a user and then a specific test needs to override it with some different behaviour

File: example-application/test/basic-tests.js [hunk 3]
```diff
@@ -95,6 +95,7 @@ describe("/api", () => {
         id: 1,
         name: "John",
       });
+      nock("http://localhost/").post(`/mailer`).reply(202);
```
[the remark this answers is not in the document]
Comment (mikicho) [reply example-application/test/basic-tests.js:98]: Totally.  I think this differentiation should be a best-practice as well!
I think a stub is a correct approach here. I'll fix it.

File: various-receipes/3rd-party-services-isolation/test/isolation.test.js (hunk 2 above)
Comment (mikicho) [reply various-receipes/3rd-party-services-isolation/test/isolation.test.js:1]: I added 3 tests:
1. shows the basic isolation approach a.k.a interceptor, nock
2. shows an unusual response like 404. (maybe I should choose 500?)
3. how to mock 3rd party services with side effects like mailer/sms to prevent spam and so on.

What cases are missing?

File: various-receipes/3rd-party-services-isolation/test/isolation.test.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+      //Assert
+      expect(orderAddResult.status).toBe(404);
+    });
+
+    test("When order failed, send mail to admin", async () => {
+      //Arrange
+      process.env.SEND_MAILS = "true";
+      sinonSandbox.stub(OrderRepository.prototype, "addOrder").throws(new Error("Unknown error"));
+      // ️️️✅ Best Practice: Intercept requests for 3rd party services to eliminate undesired side effects like emails or SMS
+      // ️️️✅ Best Practice: Specify the body when you need to make sure you call the 3rd party service as expected
```
Comment (goldbergyoni) [inline various-receipes/3rd-party-services-isolation/test/isolation.test.js:88]: I'm not 100% sure about this - How specific one should be? Should any configurable text change make the tests fail or the important part is the fact that an email was sent? Maybe the middle is a good balance - Ensure that the subject and body have a non-empty string so we know that it is a valid email but don't care about the actual test?

File: various-receipes/3rd-party-services-isolation/test/isolation.test.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+});
+
+afterAll(async (done) => {
+  // ️️️✅ Best Practice: Clean-up resources after each run
+  await stopWebServer();
+  done();
+});
+
+beforeEach(() => {
+  if (sinonSandbox) {
```
Comment (goldbergyoni) [inline various-receipes/3rd-party-services-isolation/test/isolation.test.js:37]: Should we clean nock after each call? 

Consider adding an example where there is a default answer to something (i.e. all users request) and then some specific test should override it. This triggers a challenge as the change affect the next test and might make it fail- There comes the need to cleanup after each test.

File: various-receipes/3rd-party-services-isolation/test/isolation.test.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+} = require("../../../example-application/api-under-test");
+const mailer = require("../../../example-application/libraries/mailer");
+const OrderRepository = require("../../../example-application/data-access/order-repository");
+
+let expressApp;
+let sinonSandbox;
+
+beforeAll(async (done) => {
+  // ️️️✅ Best Practice: Isolate the service under test by intercepting requests to 3rd party services 
+  nock("http://localhost/user/").get(`/1`).reply(200, {
```
Comment (goldbergyoni) [inline various-receipes/3rd-party-services-isolation/test/isolation.test.js:16]: Shouldn't we provide a default answer to ALL users call (not only to id=1)?

Comment (goldbergyoni) [inline various-receipes/3rd-party-services-isolation/test/isolation.test.js:16]: Consider blocking here all requests to collaborators as shown in #20

**Review by goldbergyoni — APPROVED** [review]
I added a few improvement comments but also approved to delegate the merge decision to you.

In overall, I think that we're doing a master degree in service isolation here...😃🚀. This will end in ~5 best practice on service isolation, absolutely unusual knowledge

It takes time because we delve into the finest details, alternatively, you may merge and later add improvement - Whatever works for you.

**Comment by goldbergyoni** [comment]
@mikicho One more comment - From now on I'll very fast to respond to this PR

File: example-application/test/basic-tests.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Thormod) [reply example-application/test/basic-tests.js:98]: @goldbergyoni Can you elaborate more your idea? as teams usually use `nock` is for intercepting 3rd parties calls. The way we have used among projects is recording this calls and use it as fixtures in the next execution.

[the remark this answers is not in the document]
Comment (goldbergyoni) [reply example-application/test/basic-tests.js:101]: @Thormod

Differentiate between two scenarios:
1. A:Our code makes HTTP call - Our Order Microservice is calling the User Microservice, we are aware of this call and it is unlikely to change without our notice
2. B: Other code is making HTTP call - We use some npm lib, like AWS SES, that under the good is making some HTTP call. This might change without our notice

To my understanding, A is valid for knocking, B not

File: various-receipes/3rd-party-services-isolation/test/isolation.test.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+const OrderRepository = require("../../../example-application/data-access/order-repository");
+
+let expressApp;
+let sinonSandbox;
+
+beforeAll(async (done) => {
+  // ️️️✅ Best Practice: Place the backend under test within the same process
+  expressApp = await initializeWebServer();
+  // ️️️✅ Best Practice: Ensure that this component is isolated by preventing unknown calls except for the Api-Under-Test
+  nock.disableNetConnect();
```
Comment (goldbergyoni) [inline various-receipes/3rd-party-services-isolation/test/isolation.test.js:17]: Is disable needed? Wouldn't the next line, enableNetConnect, block all non-allowed requests?

File: various-receipes/3rd-party-services-isolation/test/isolation.test.js [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+  })
+
+  if (sinonSandbox) {
+    sinonSandbox.restore();
+  }
+});
+
+afterEach(() => {
+  // ️️️✅ Best Practice: Clean nock interceptors between tests
+  nock.cleanAll();
```
Comment (goldbergyoni) [inline various-receipes/3rd-party-services-isolation/test/isolation.test.js:46]: Just curious - Why did you prefer afterEach over beforeEach?

**Review by goldbergyoni — APPROVED** [review]
Looks great to me!

File: various-receipes/3rd-party-services-isolation/test/isolation.test.js (hunk 7 above)
Comment (mikicho) [reply various-receipes/3rd-party-services-isolation/test/isolation.test.js:17]: nope, see [docs](https://github.com/nock/nock#enabling-requests)

File: various-receipes/3rd-party-services-isolation/test/isolation.test.js (hunk 8 above)
Comment (mikicho) [reply various-receipes/3rd-party-services-isolation/test/isolation.test.js:46]: I don't know.. clean some something of `afterEach` isn't it?
